include/mbgl/storage/response.hpp

```diff
@@ -35,7 +35,7 @@ class Response {
     optional<std::string> etag;
 
     bool isFresh() const {
-        return !expires || *expires > util::now();
+        return (!expires || *expires > util::now()) && !error;
```

Comment:
> Should this condition be expires ? *expires > util::now() : !error? In case there is an explicit Cache-Control or Expires header on an error response, it should be respected.
> 

I think they are logically the same. But yours is definitely easier to read.

> Also, error can be Reason::NotFound (404). We should carefully consider (and test) what the retry behavior is in that situation.

Currently NotFound doesn't trigger a retry which makes sense, specially for tiles. I can't think of a use case we should retry for styles.

I added 2 more tests to make sure this is what we want.